janusgraph-core/src/main/java/org/janusgraph/util/datastructures/SetUtil.java

```diff
@@ -0,0 +1,51 @@
+// Copyright 2020 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.util.datastructures;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * @author Florian Grieskamp (Florian.Grieskamp@gdata.de)
+ */
+public class SetUtil {
+    
+    public static <T> Set<Set<T>> powerSet(final Set<T> originalSet) {
+        int resultSize = (int) Math.pow(2, originalSet.size());
+        Set<Set<T>> resultSets = new HashSet<Set<T>>(resultSize);
+
+        resultSets.add(new HashSet<T>(0));
+
+        if (originalSet.isEmpty()) {
+            return resultSets;
+        }
+
+        Set<T> remaining = new HashSet<T>(originalSet);
+        T first = remaining.iterator().next();
+        remaining.remove(first);
+
+        for (Set<T> intermediateSet : powerSet(remaining)) {
```

Comment:
Did you explore Guava's PowerSet implementation? Their implementation seems to be more efficient: https://www.baeldung.com/java-power-set-of-a-set#optimization, and more importantly, it only uses O(n) space by utilizing the Iterator interface.

Reply:
I refactored the code, it should not exceed the O(n) space bound now